### DISCUSSION TITLE
PrismaClientInitializationError on Next.js app router

### DISCUSSION THREAD

**Discussion by edtan094** [prisma/prisma#24280] [Q&A]
### Question

Hello, 

I can't figure out why this error is occurring only in production on my next.js v14 app router.  This error and other similar errors occur throughout my application.  The only place it does not break in my application is on my home page for some reason(perhaps because the information shown on the home page is static?).    Below are a few examples of the errors I have encountered.  

PrismaClientInitializationError: 
Invalid `prisma.order.aggregate()` invocation:

```
Error querying the database: Error code 14: Unable to open the database file
    at In.handleRequestError (/var/task/node_modules/@prisma/client/runtime/library.js:122:7154)
    at In.handleAndLogRequestError (/var/task/node_modules/@prisma/client/runtime/library.js:122:6188)
    at In.request (/var/task/node_modules/@prisma/client/runtime/library.js:122:5896)
    at async l (/var/task/node_modules/@prisma/client/runtime/library.js:127:11167)
    at async p.revalidate (/var/task/.next/server/app/admin/page.js:11:1875)
    at async /var/task/.next/server/chunks/708.js:1:5707
    at async Promise.all (index 0)
    at async f (/var/task/.next/server/app/admin/page.js:11:2573) {
  clientVersion: '5.12.1',
  errorCode: undefined,
  digest: '3119279550'
}

PrismaClientInitializationError: 
Invalid `prisma.product.findUnique()` invocation:
```
```
Error querying the database: Error code 14: Unable to open the database file
    at In.handleRequestError (/var/task/node_modules/@prisma/client/runtime/library.js:122:7154)
    at In.handleAndLogRequestError (/var/task/node_modules/@prisma/client/runtime/library.js:122:6188)
    at In.request (/var/task/node_modules/@prisma/client/runtime/library.js:122:5896)
    at async l (/var/task/node_modules/@prisma/client/runtime/library.js:127:11167)
    at async p (/var/task/.next/server/app/(customerFacing)/products/[id]/purchase/page.js:1:15416) {
  clientVersion: '5.12.1',
  errorCode: undefined,
  digest: '2796859334'
}
```

### Information about Prisma Schema, Client Queries and Environment (optional)

Here is my schema file

```prisma
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "sqlite"
  url      = env("DATABASE_URL")
}

model Product {
  id                     String                 @id @default(uuid())
  name                   String
  priceInCents           Int
  filePath               String
  imagePath              String
  description            String
  isAvailableForPurchase Boolean                @default(true)
  createdAt              DateTime               @default(now())
  updatedAt              DateTime               @updatedAt
  orders                 Order[]
  downloadVerifications  DownloadVerification[]
}

model User {
  id        String   @id @default(uuid())
  email     String   @unique
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  orders    Order[]
}

model Order {
  id               String   @id @default(uuid())
  pricePaidInCents Int
  createdAt        DateTime @default(now())
  updatedAt        DateTime @updatedAt

  userId    String
  productId String
  user      User    @relation(fields: [userId], references: [id], onDelete: Cascade)
  product   Product @relation(fields: [productId], references: [id], onDelete: Restrict)
}

model DownloadVerification {
  id        String   @id @default(uuid())
  expiresAt DateTime
  createdAt DateTime @default(now())
  productId String
  product   Product  @relation(fields: [productId], references: [id], onDelete: Cascade)
}
```

The file name of the db is dev.db which matches the env variable placed in Vercel of file:./dev.db in accordance to the Prisma deployment documentation.  

Here are some of my queries

```
async function getSalesData() {
  const data = await db.order.aggregate({
    _sum: { pricePaidInCents: true },
    _count: true,
  });

  return {
    amount: (data._sum.pricePaidInCents || 0) / 100,
    numberOfSales: data._count,
  };
}

async function getUserData() {
  const [userCount, orderData] = await Promise.all([
    db.user.count(),
    db.order.aggregate({
      _sum: { pricePaidInCents: true },
    }),
  ]);

  return {
    userCount,
    averageValuePerUser:
      userCount === 0
        ? 0
        : (orderData._sum.pricePaidInCents || 0) / userCount / 100,
  };
}

async function getProductData() {
  const [activeCount, inactiveCount] = await Promise.all([
    db.product.count({ where: { isAvailableForPurchase: true } }),
    db.product.count({ where: { isAvailableForPurchase: false } }),
  ]);

  return { activeCount, inactiveCount };
}
```

- OS: 
- Database: SQLite
- Node.js version:  Node 18

**Comment by ludralph** [comment]
Hi @edtan094 👋 

Thank you for raising this question.

The errors you are encountering in production with your Next.js application indicate issues related to database file access. Specifically, the Error querying the database: `Error code 14: Unable to open the database file` suggests that your SQLite database file dev.db is not accessible in the production environment. 

[Vercel](https://vercel.com/guides/is-sqlite-supported-in-vercel) does not have support for SQLite but suggest using [Vercel-postgres](https://vercel.com/storage/postgres). You can take a look at the linked documentation to see how to transition to Vercel postgres. If you do want to keep using SQLite, then you would need to use another platform that has support for SQLite.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
